tt_eager/tt_dnn/op_library/rotary_embedding/rotary_embedding_llama_op.hpp

```diff
@@ -0,0 +1,95 @@
+
+    operation::ProgramWithCallbacks create_program(
+        const std::vector<Tensor> &input_tensors, std::vector<Tensor> &output_tensors) const;
+    tt::stl::reflection::Attributes attributes() const;
+
+    const operation::Hash compute_program_hash(const std::vector<Tensor> &input_tensors) const;
+};
+
+inline Tensor rotary_embedding_llama(
+    const Tensor &input_tensor,
+    const Tensor &cos,
+    const Tensor &sin,
+    const Tensor trans_mat,
+    const MemoryConfig &output_mem_config = operation::DEFAULT_OUTPUT_MEMORY_CONFIG,
+    std::optional<const DeviceComputeKernelConfig> compute_kernel_config = std::nullopt) {
+    std::vector<Tensor> output_tensors = {Tensor(operation::get_workers_for_op_output({input_tensor, cos, sin, trans_mat}))};
+    operation::launch_with_autoformat(
+        [output_mem_config, compute_kernel_config] (const std::vector<Tensor>& input_tensors, const std::vector<std::optional<const Tensor>>& optional_input_tensors, const std::vector<std::optional<Tensor>>& optional_output_tensors) mutable -> std::vector<Tensor> {
+            auto& input_tensor = input_tensors.at(0);
+            auto& cos = input_tensors.at(1);
+            auto& sin = input_tensors.at(2);
+            auto& trans_mat = input_tensors.at(3);
+            TT_FATAL(input_tensor.get_legacy_shape()[-1] % TILE_WIDTH == 0, "Input X dim must be divisible into tiles");
+            uint32_t seq_len = input_tensor.get_legacy_shape()[-2];
+            uint32_t B = input_tensor.get_legacy_shape()[0];
+            uint32_t head_dim = input_tensor.get_legacy_shape()[-1];
+            TT_FATAL(cos.get_legacy_shape() == sin.get_legacy_shape(), "Cos and Sin dims must match");
+            TT_FATAL(cos.get_legacy_shape()[0] == 1 && cos.get_legacy_shape()[1] == 1 && cos.get_legacy_shape()[-1] == head_dim, "Cos dims must match input dims");
+            TT_FATAL(cos.get_legacy_shape()[-2] >= seq_len, "Cos dims must match input dims");
+
+            TT_FATAL(trans_mat.get_legacy_shape()[0] == 1 && trans_mat.get_legacy_shape()[1] == 1, "Transformation matrix must have 1st & 2nd dim equal to 1");
+            TT_FATAL(trans_mat.get_legacy_shape()[-2] == TILE_HEIGHT, "Transformation matrix must have 3rd dim equal to TILE_HEIGHT");
+            TT_FATAL(trans_mat.get_legacy_shape()[-1] == TILE_WIDTH, "Transformation matrix must have 4rd dim equal to TILE_WIDTH");
+
+            auto arch = input_tensor.storage_type() == StorageType::DEVICE ? input_tensor.device()->arch() : AutoFormat::GetDefaultDevice()->arch();
+            auto kernel_config_val = init_device_compute_kernel_config(arch, compute_kernel_config, MathFidelity::HiFi4, true, false, false);
+
+            TT_FATAL(head_dim <= 128 || std::get<WormholeComputeKernelConfig>(compute_kernel_config.value()).fp32_dest_acc_en == false, "If head_dim is > 128, fp32_dest_acc_en must be False");
+
+            Shape input_pad_shape = AutoFormat::pad_to_tile_shape(input_tensor.get_legacy_shape());
```

Comment:
@TT-BrianLiu do you think there is a good reason to run with autoformat here? I'm not familiar with the difference between this and operation::run, but it seems overly complicated for this op.

Reply:
We are moving towards removing run_with_autoformat. Would be good to remove here if possible. User should format the input to the op's requirements using available methods (for ttnn can use to_layout)

Reply:
Yes, agreed. We are going to move away from autoformatting in ops